artemis-commons/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQScheduledComponent.java

```diff
@@ -287,35 +321,63 @@ private void restartIfNeeded() {
       }
    }
 
-   final Runnable runForExecutor = new Runnable() {
-      @Override
-      public void run() {
-         if (onDemand && delayed.get() > 0) {
-            delayed.decrementAndGet();
+   private void runForExecutor(AtomicBoolean bookedForRunning) {
+      if (bookedForRunning != null) {
+         if (!bookedForRunning.compareAndSet(true, false)) {
+            // this can happen only if there is a racing stop
+            return;
          }
+      }
+      ActiveMQScheduledComponent.this.run();
+   }
 
-         if (!onDemand && lastTime > 0) {
-            if (System.currentTimeMillis() - lastTime < millisecondsPeriod) {
-               logger.trace("Execution ignored due to too many simultaneous executions, probably a previous delayed execution");
+   private void delayedRunForScheduler(AtomicBoolean bookedForRunning) {
+      if (executor != null) {
+         if (bookedForRunning != null) {
+            if (!bookedForRunning.get()) {
+               // this can happen only if there is a racing stop
                return;
             }
          }
-
-         lastTime = System.currentTimeMillis();
-
-         ActiveMQScheduledComponent.this.run();
+         try {
+            executor.execute(() -> runForExecutor(bookedForRunning));
+         } catch (RejectedExecutionException e) {
+            if (bookedForRunning != null) {
+               bookedForRunning.set(false);
+            }
+            throw e;
+         }
+      } else {
+         runForExecutor(bookedForRunning);
       }
-   };
-
-   final Runnable runForScheduler = new Runnable() {
-      @Override
-      public void run() {
-         if (executor != null) {
-            executor.execute(runForExecutor);
-         } else {
-            runForExecutor.run();
+   }
+
+   private void runForScheduler(AtomicBoolean bookedForRunning) {
+      if (executor != null) {
+         if (bookedForRunning != null && !bookedForRunning.compareAndSet(false, true)) {
```

Comment:
Both of the 'if executor != null' legs do this check first, I think it would be simpler to move it outside the if and do it once atthe start. The descriptions dont differ enough to justify both IMO, so its more rather than less complicated/confusing having both.